controllers/ingress_controller.go

```diff
@@ -80,102 +116,142 @@ func (r *IngressReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 
 	// /FINALIZER LOGIC ================================================================================================
 
-	if desired.Spec.IngressClassName == nil {
-		log.Info("IngressClassName should not be nil")
-		return ctrl.Result{}, errors.New("IngressClassName should not be nil")
+	// ensure there is a template api definition
+	templateAnnotationValue, ok := desired.Annotations[ingressTemplateAnnotationKey]
+	if !ok {
+		return ctrl.Result{}, fmt.Errorf("expecting template annotation %s", ingressTemplateAnnotationKey)
 	}
 
-	ingressClassList := &v1beta1.IngressClassList{}
-	opts := []client.ListOption{
-		client.InNamespace(req.Namespace),
-	}
-	err := r.List(ctx, ingressClassList, opts...)
+	// we have parameters - as such, we should ensure that there is an api definition resource
+	template := &v1alpha1.ApiDefinition{}
+	err := r.Get(ctx, types.NamespacedName{Name: templateAnnotationValue, Namespace: req.Namespace}, template)
 	if err != nil {
-		println(err.Error())
+		log.Error(err, "error getting api definition to use as a template", "name", templateAnnotationValue)
 		return ctrl.Result{}, err
 	}
 
-	println(ingressClassList.Kind)
+	apisToCreateOrUpdate := v1alpha1.ApiDefinitionList{}
+	apisToUpdate := v1alpha1.ApiDefinitionList{}
+	apisToCreate := v1alpha1.ApiDefinitionList{}
+	apisToDelete := v1alpha1.ApiDefinitionList{}
 
-	ingressClass := &v1beta1.IngressClass{}
-	if err := r.Get(ctx, types.NamespacedName{Name: *desired.Spec.IngressClassName, Namespace: req.Namespace}, ingressClass); err != nil {
-		println(err.Error())
-		return ctrl.Result{}, err
-	}
+	for _, rule := range desired.Spec.Rules {
+		hostName := rule.Host
+		for _, p := range rule.HTTP.Paths {
+			apiName := strings.TrimRight(strings.ReplaceAll(fmt.Sprintf("%s-%s-%s-%s", req.Name, req.Namespace, hostName, p.Path), "/", ""), "-")
+			api := v1alpha1.ApiDefinition{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      apiName,
+					Namespace: req.Namespace,
+					Labels:    lbls,
+				},
+				Spec: template.Spec,
+			}
+
+			api.Spec.Name = apiName
+			api.Spec.Proxy.ListenPath = p.Path
+			api.Spec.Proxy.TargetURL = fmt.Sprintf("http://%s.%s.svc.cluster.local:%d", p.Backend.ServiceName, namespacedName.Namespace, p.Backend.ServicePort.IntValue())
```

Comment:
What does this do? In tyk perspective, Target actually means upstream. This is user provided, we shouldn't be doing anything on this field. This will be something like `http://httpbin.org` now if they want to target cluster services they should explicitly state it and dns will resolve accordingly.

Reply:
The whole purpose of ingress is to target cluster services

Reply:
>The whole purpose of ingress is to target cluster services?

Yes, and in our case they are `api's` behind  tyk in the cluster.

we are `Incomming traffic` `=>` `tyk/api-service-endpoint`

Reply:
we are exposing the api's to the outside world. This is what I think.

Reply:
@gernest after catchup - is it ok to resolve this conversation?